spv/sync.go

```diff
@@ -0,0 +1,1420 @@
+// Copyright (c) 2018-2019 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package spv
+
+import (
+	"context"
+	"net"
+	"runtime"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/decred/dcrd/addrmgr"
+	"github.com/decred/dcrd/chaincfg/chainhash"
+	"github.com/decred/dcrd/chaincfg/v2"
+	"github.com/decred/dcrd/gcs/blockcf"
+	"github.com/decred/dcrd/wire"
+	"github.com/decred/dcrwallet/errors/v2"
+	"github.com/decred/dcrwallet/lru"
+	"github.com/decred/dcrwallet/p2p/v2"
+	"github.com/decred/dcrwallet/validate"
+	"github.com/decred/dcrwallet/wallet/v3"
+	"golang.org/x/sync/errgroup"
+)
+
+// reqSvcs defines the services that must be supported by outbounded peers.
+// After fetching more addresses (if needed), peers are disconnected from if
+// they do not provide each of these services.
+const reqSvcs = wire.SFNodeNetwork | wire.SFNodeCF
+
+// Syncer implements wallet synchronization services by over the Decred wire
+// protocol using Simplified Payment Verification (SPV) with compact filters.
+type Syncer struct {
+	// atomics
+	atomicCatchUpTryLock uint32 // CAS (entered=1) to perform discovery/rescan
+
+	rescanningWalletID int
+	wallets            map[int]*wallet.Wallet
+	chainParams        *chaincfg.Params
+	lp                 *p2p.LocalPeer
+
+	// Protected by atomicCatchUpTryLock
+	discoverAccountsIfUnlocked bool
+	loadedFilters              map[int]bool
+
+	persistentPeers []string
+
+	connectingRemotes map[string]struct{}
+	remotes           map[string]*p2p.RemotePeer
+	remotesMu         sync.Mutex
+
+	// Data filters
+	//
+	// TODO: Replace precise rescan filter with wallet db accesses to avoid
+	// needing to keep all relevant data in memory.
+	rescanFilter   map[int]*wallet.RescanFilter
+	filterData     map[int]*blockcf.Entries
```

Comment:
This wasn't a pointer to `blockcf.Entries`. Any reason for this change to pointer?

Reply:
Being stored in a map requires it to be a pointer in order to call pointer functions.